File: sdks/kotlin/src/test/kotlin/dev/wheso/TransportTest.kt
// 疎通試験（Kotlin、段 B）。
//
// 何を証明するか: Kotlin の SDK が**実際の Durable Object**（partykit dev）へ WebSocket で
// 接続し、実際に符号化された AV1（spec/vectors/real-media.json）を送り、購読者として
// **1 バイトも変わらずに**受け取れること。段 A（凍結ベクタとトレース）は判断の同一性を
// 示すが、実際に線を通ることは示さない。
//
// 依存を追加しない: WebSocket は java.net.http（JDK 11 以降）、HMAC は javax.crypto を使う。
//
// 実行の前提: 局所実行環境が起動していること。環境変数で場所と鍵を受け取る
// （WHESO_WS_BASE / WHESO_ROOM / WHESO_NODE_KEY / WHESO_SENDER_PK / WHESO_SUB_PK）。
// 無い場合は飛ばす。起動は tools/transport-suite.ts の責務である。
package dev.wheso

import java.io.File
import java.net.URI
import java.net.http.HttpClient
import java.net.http.WebSocket
import java.nio.ByteBuffer
import java.util.Base64
import java.util.concurrent.CompletionStage
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class TransportTest {
    /** 時刻窓の長さ。認証規範の NODE_AUTH_TIME_WINDOW_SEC と一致させる。 */
    private val timeWindowSec = 300L

    /** 送る枚数。全部送ると試験が長くなるため先頭に限る（キーフレームを含む）。 */
    private val sendCount = 10

    private fun hmacSha256(key: ByteArray, message: String): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(message.toByteArray(Charsets.UTF_8))
    }

    /** base64url（詰め物なし）。認証規範の符号化に合わせる。 */
    private fun base64Url(bytes: ByteArray): String =
        Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

    /**
     * nodeHello の authTag を作る。
     * 会議シークレット = HMAC(鍵, "meeting-secret:v1:<会議 ID>")
     * authTag = base64url(HMAC(会議シークレット, "node-auth:v1:<部屋名>:<役割>:<時刻窓>"))
     * 参照実装（packages/core/src/auth.ts）と同じ文字列でなければ接続が拒否される。
     */
    private fun buildAuthTag(key: String, room: String, role: String): String {
        val meetingId = room.split("-").getOrNull(1) ?: ""
        val secret = hmacSha256(key.toByteArray(Charsets.UTF_8), "meeting-secret:v1:$meetingId")
        val window = (System.currentTimeMillis() / 1000L) / timeWindowSec
        return base64Url(hmacSha256(secret, "node-auth:v1:$room:$role:$window"))
    }

    private fun hexToBytes(hex: String): ByteArray {
        val bytes = ByteArray(hex.length / 2)
        for (index in bytes.indices) {
            bytes[index] = hex.substring(index * 2, index * 2 + 2).toInt(16).toByte()
        }
        return bytes
    }

    private fun bytesToHex(bytes: ByteArray): String {
        val builder = StringBuilder(bytes.size * 2)
        for (byte in bytes) {
            builder.append("%02x".format(byte.toInt() and 0xFF))
        }
        return builder.toString()
    }

    private fun readAsset(): JsonObject =
        Json.parseToJsonElement(File("../../spec/vectors/real-media.json").readText()).jsonObject

    /** 受信したバイナリを溜める購読者。断片が来る場合があるため last まで繋ぐ。 */
    private class Collector(private val expected: Int) : WebSocket.Listener {
        val received = ConcurrentLinkedQueue<ByteArray>()
        val done = CountDownLatch(1)
        private val partial = mutableListOf<Byte>()

        override fun onOpen(webSocket: WebSocket) {
            webSocket.request(Long.MAX_VALUE)
        }

        override fun onBinary(webSocket: WebSocket, data: ByteBuffer, last: Boolean): CompletionStage<*>? {
            val chunk = ByteArray(data.remaining())
            data.get(chunk)
            partial.addAll(chunk.toList())
            if (last) {
                received.add(partial.toByteArray())
                partial.clear()
                if (received.size >= expected) {
                    done.countDown()
                }
            }
            webSocket.request(1)
            return null
        }

        override fun onClose(webSocket: WebSocket, statusCode: Int, reason: String): CompletionStage<*>? {
            println("購読側が閉じられた: $statusCode $reason")
            done.countDown()
            return null
        }

        override fun onError(webSocket: WebSocket, error: Throwable) {
            println("購読側の誤り: ${error.message}")
            done.countDown()
        }
    }

    private fun connect(url: String, listener: WebSocket.Listener): WebSocket =
        HttpClient.newHttpClient()
            .newWebSocketBuilder()
            .buildAsync(URI.create(url), listener)
            .get(30, TimeUnit.SECONDS)

    @Test
    fun realMediaTravelsThroughLiveNode() {
        val environment = System.getenv()
        val base = environment["WHESO_WS_BASE"]
        val room = environment["WHESO_ROOM"]
        val key = environment["WHESO_NODE_KEY"]
        val senderPk = environment["WHESO_SENDER_PK"]
        val subPk = environment["WHESO_SUB_PK"]
        if (base == null || room == null || key == null || senderPk == null || subPk == null) {
            // 局所実行環境が無い場所では飛ばす。CI では実行器が環境変数を与える。
            println("SKIP 疎通試験（環境変数が無い）")
            return
        }

        val asset = readAsset()
        val video = asset["video"]?.jsonObject ?: error("video が無い")
        val frames = video["frames"]?.jsonArray ?: error("frames が無い")
        val senderId = asset["senderId"]?.jsonPrimitive?.content?.toLong() ?: error("senderId が無い")
        val channel = video["channel"]?.jsonPrimitive?.content?.toInt() ?: error("channel が無い")
        val framerate = video["framerate"]?.jsonPrimitive?.content?.toInt() ?: 30
        assertTrue(frames.size >= sendCount, "資産に $sendCount 枚以上ある")

        // 購読側を先に開く。順序を逆にすると転送先が無く、送ったものが消える。
        val collector = Collector(sendCount)
        val subscriber = connect("$base/parties/shard/$room?_pk=$subPk", collector)
        // JDK の WebSocket は前の送信が完了する前に次を呼ぶと失敗する。1 件ずつ待つ。
        subscriber.sendText(
            """{"t":"nodeHello","role":"receiver","nodeId":"$room","authTag":"${buildAuthTag(key, room, "receiver")}"}""",
            true,
        ).get(10, TimeUnit.SECONDS)
        subscriber.sendText(
            """{"t":"subscribe","entries":[{"senderId":$senderId,"channel":$channel,"maxSpatialId":3,"maxTemporalId":7}]}""",
            true,
        ).get(10, TimeUnit.SECONDS)
        // 購読の登録が処理される猶予を置く。
        Thread.sleep(1000)

        val sender = connect(
            "$base/parties/shard/$room?_pk=$senderPk",
            object : WebSocket.Listener {
                override fun onOpen(webSocket: WebSocket) {
                    webSocket.request(Long.MAX_VALUE)
                }

                override fun onClose(webSocket: WebSocket, statusCode: Int, reason: String): CompletionStage<*>? {
                    println("送信側が閉じられた: $statusCode $reason")
                    return null
                }
            },
        )
        sender.sendText(
            """{"t":"nodeHello","role":"sender","nodeId":"$room","authTag":"${buildAuthTag(key, room, "sender")}"}""",
            true,
        ).get(10, TimeUnit.SECONDS)
        Thread.sleep(500)

        val sentHex = mutableListOf<String>()
        for (index in 0 until sendCount) {
            val frame = frames[index].jsonObject
            val hex = frame["expectedMessageHex"]?.jsonPrimitive?.content ?: error("期待バイト列が無い")
            sentHex.add(hex)
            sender.sendBinary(ByteBuffer.wrap(hexToBytes(hex)), true).get(10, TimeUnit.SECONDS)
            // 実際の間隔で送る。詰めて送ると予算超過の破棄が働き、疎通の検証にならない。
            Thread.sleep((1000L / framerate).coerceAtLeast(1L))
        }

        assertTrue(
            collector.done.await(30, TimeUnit.SECONDS),
            "購読者が $sendCount 件を受け取る（実際 ${collector.received.size} 件）",
        )
        val receivedHex = collector.received.map { bytesToHex(it) }
        assertEquals(sendCount, receivedHex.size, "受け取った件数が一致する")
        for (index in 0 until sendCount) {
            assertEquals(
                sentHex[index],
                receivedHex[index],
                "$index 番目のバイト列が 1 バイトも変わらない",
            )
        }

        subscriber.abort()
        sender.abort()
    }
}

File: sdks/dart/test/transport_test.dart
// 疎通試験（Dart、段 B）。
//
// 何を証明するか: Dart の SDK が**実際の Durable Object**（partykit dev）へ WebSocket で
// 接続し、実際に符号化された AV1（spec/vectors/real-media.json）を送り、購読者として
// **1 バイトも変わらずに**受け取れること。段 A（凍結ベクタ・トレース・実データ照合）は
// 判断とバイト列の同一性を示すが、実際に線を通ることは示さない。
//
// 依存を追加しない: WebSocket は dart:io（標準）、HMAC は test/support/digest.dart の
// 自前実装を使う。自前 HMAC は digest_test.dart が RFC 4231 で検証している。
//
// 実行の前提: 実環境（PartyKit managed）へデプロイされていること。環境変数で場所と鍵を受け取る
// （WHESO_WS_BASE / WHESO_ROOM / WHESO_NODE_KEY / WHESO_SENDER_PK / WHESO_SUB_PK）。
// 無い場合は飛ばす。起動は tools/transport-suite.ts の責務である。

import 'dart:convert';
import 'dart:io';

import 'package:test/test.dart';

import 'support/digest.dart';

/// 時刻窓の長さ。認証規範の NODE_AUTH_TIME_WINDOW_SEC と一致させる。
const int timeWindowSec = 300;

/// 送る枚数。全部送ると試験が長くなるため先頭に限る（キーフレームを含む）。
const int sendCount = 10;

/// nodeHello の authTag を作る。
/// 会議シークレット = HMAC(鍵, "meeting-secret:v1:<会議 ID>")
/// authTag = base64url(HMAC(会議シークレット, "node-auth:v1:<部屋名>:<役割>:<時刻窓>"))
/// 参照実装（packages/core/src/auth.ts）と 1 文字でも違えば 4023 で切られる。
String buildAuthTag(String key, String room, String role) {
  final List<String> parts = room.split('-');
  final String meetingId = parts.length > 1 ? parts[1] : '';
  final List<int> secret =
      hmacSha256(utf8.encode(key), utf8.encode('meeting-secret:v1:$meetingId'));
  final int window =
      (DateTime.now().millisecondsSinceEpoch ~/ 1000) ~/ timeWindowSec;
  return base64UrlNoPad(
    hmacSha256(secret, utf8.encode('node-auth:v1:$room:$role:$window')),
  );
}

/// 資産を読む。試験の作業ディレクトリは sdks/dart である。
Object? readAsset() {
  final File file = File('../../spec/vectors/real-media.json');
  return jsonDecode(file.readAsStringSync());
}

/// 型を確かめて地図として取り出す。dynamic を使わない。
Map<String, Object?> asMap(Object? value, String what) {
  if (value is Map<String, Object?>) {
    return value;
  }
  throw StateError('$what が地図でない');
}

List<Object?> asList(Object? value, String what) {
  if (value is List<Object?>) {
    return value;
  }
  throw StateError('$what が並びでない');
}

int asInt(Object? value, String what) {
  if (value is int) {
    return value;
  }
  throw StateError('$what が整数でない');
}

String asText(Object? value, String what) {
  if (value is String) {
    return value;
  }
  throw StateError('$what が文字列でない');
}

void main() {
  test('実データが実行中のノードを通って戻る', () async {
    final Map<String, String> environment = Platform.environment;
    // dart:io の WebSocket は TLS を話せるため、実環境へ直に繋ぐ（wss）。
    // WHESO_WSS_BASE が無い場合だけ平文の口（TLS 終端）へ落ちる。
    final String? base =
        environment['WHESO_WSS_BASE'] ?? environment['WHESO_WS_BASE'];
    final String? room = environment['WHESO_ROOM'];
    final String? key = environment['WHESO_NODE_KEY'];
    final String? senderPk = environment['WHESO_SENDER_PK'];
    final String? subPk = environment['WHESO_SUB_PK'];
    if (base == null ||
        room == null ||
        key == null ||
        senderPk == null ||
        subPk == null) {
      // 局所実行環境が無い場所では飛ばす。実行器が環境変数を与える。
      print('SKIP 疎通試験（環境変数が無い）');
      return;
    }

    final Map<String, Object?> asset = asMap(readAsset(), '資産');
    final Map<String, Object?> video = asMap(asset['video'], 'video');
    final List<Object?> frames = asList(video['frames'], 'frames');
    final int senderId = asInt(asset['senderId'], 'senderId');
    final int channel = asInt(video['channel'], 'channel');
    final Object? rawFramerate = video['framerate'];
    final int framerate = rawFramerate is int ? rawFramerate : 30;
    expect(frames.length, greaterThanOrEqualTo(sendCount));

    // 購読側を先に開く。順序を逆にすると転送先が無く、送ったものが消える。
    final WebSocket subscriber =
        await WebSocket.connect('$base/parties/shard/$room?_pk=$subPk');
    final List<String> receivedHex = <String>[];
    // 受信を先に仕掛ける。仕掛ける前に送ると取りこぼす。
    subscriber.listen(
      (Object? message) {
        if (message is List<int>) {
          receivedHex.add(bytesToHex(message));
        }
        // 制御メッセージ（nodeHelloAck など）は数えない。
      },
      onDone: () {
        // 認証に失敗すると 4023（E_NODE_AUTH）で切られる。原因を残す。
        print('購読側が閉じられた: ${subscriber.closeCode} ${subscriber.closeReason}');
      },
    );

    subscriber.add(
      '{"t":"nodeHello","role":"receiver","nodeId":"$room","authTag":"${buildAuthTag(key, room, 'receiver')}"}',
    );
    subscriber.add(
      '{"t":"subscribe","entries":[{"senderId":$senderId,"channel":$channel,"maxSpatialId":3,"maxTemporalId":7}]}',
    );
    // 購読の登録が処理される猶予を置く。
    await Future<void>.delayed(const Duration(milliseconds: 1000));

    final WebSocket sender =
        await WebSocket.connect('$base/parties/shard/$room?_pk=$senderPk');
    sender.listen(
      (Object? message) {},
      onDone: () {
        print('送信側が閉じられた: ${sender.closeCode} ${sender.closeReason}');
      },
    );
    sender.add(
      '{"t":"nodeHello","role":"sender","nodeId":"$room","authTag":"${buildAuthTag(key, room, 'sender')}"}',
    );
    await Future<void>.delayed(const Duration(milliseconds: 500));

    final List<String> sentHex = <String>[];
    for (int index = 0; index < sendCount; index++) {
      final Map<String, Object?> frame = asMap(frames[index], 'frame');
      final String hex = asText(frame['expectedMessageHex'], 'expectedMessageHex');
      sentHex.add(hex);
      sender.add(hexToBytes(hex));
      // 実際の間隔で送る。詰めて送ると予算超過の破棄が働き、疎通の検証にならない。
      await Future<void>.delayed(
        Duration(milliseconds: (1000 / (framerate < 1 ? 1 : framerate)).ceil()),
      );
    }

    // 転送は非同期であるため、送り終えた時点では届いていない。揃うまで待つ。
    final DateTime deadline = DateTime.now().add(const Duration(seconds: 30));
    while (receivedHex.length < sendCount && DateTime.now().isBefore(deadline)) {
      await Future<void>.delayed(const Duration(milliseconds: 50));
    }

    expect(receivedHex.length, sendCount, reason: '購読者が $sendCount 件を受け取る');
    for (int index = 0; index < sendCount; index++) {
      expect(
        receivedHex[index],
        sentHex[index],
        reason: '$index 番目のバイト列が 1 バイトも変わらない',
      );
    }

    await subscriber.close();
    await sender.close();
  }, timeout: const Timeout(Duration(minutes: 2)));
}
